src/cs50/sql.py

```diff
@@ -232,6 +258,38 @@ def execute(self, sql, *args, **kwargs):
         # Join tokens into statement
         statement = "".join([str(token) for token in tokens])
 
+        # Connect to database (for transactions' sake)
+        try:
+
+            # Infer whether Flask is installed
+            import flask
+
+            # Infer whether app is defined
+            assert flask.current_app
+
+            # If no connection for app's current request yet
+            if not hasattr(flask.g, "_connection"):
+
+                # Connect now
+                flask.g._connection = self._engine.connect()
+
+                # Disconnect later
+                @flask.current_app.teardown_appcontext
+                def shutdown_session(exception=None):
+                    flask.g._connection.close()
+
+            # Use this connection
+            connection = flask.g._connection
+
+        except (ModuleNotFoundError, AssertionError):
+
+            # If no connection yet
+            if not hasattr(self, "_connection"):
+                self._connection = self._engine.connect()
```

Comment:
@dmalan would this connection be leaked if the user uses `db.execute` at least once outside of an app context and then uses it again inside one?

Reply:
Or I guess should we close this connection at some point?

Reply:
I suspect it's garbage-collected, but added

```
    def __del__(self):
        """Close database connection."""
        if hasattr(self, "_connection"):
            self._connection.close()
```

for good measure.